armTemplates/azureDeploy.json

```diff
@@ -0,0 +1,843 @@
+            }
+        },
+        "vmAdminPassword": {
+            "type": "securestring",
+            "metadata": {
+                "description": "The password to use to manage to the virtual machine hosting the web server."
+            }
+        },
+        "vmAdminUsername": {
+            "type": "securestring",
+            "metadata": {
+                "description": "The username to use to manage to the virtual machine hosting the web server."
+            }
+        }
+
+    },
+    "functions":[
+        {
+            "namespace": "custom",
+            "members": {
+                "cleanName": {
+                    "parameters": [
+                        {
+                        "name": "dirtyName",
+                        "type": "string"
+                        }
+                    ],
+                    "output": {
+                        "type": "string",
+                        "value": "[replace(replace(replace(parameters('dirtyName'), '-', ''), '_', ''), ' ', '')]"
+                    }
+                }
+            }
+        }
+    ],
+    "variables": {
+        "app_networkSecurityGroup_name": "[concat(parameters('projectName'), '-APP-NSG')]",
+        "app_subnet_addressPrefix": "[concat(parameters('virtualNetworkAddressBase'), '.32/27')]",
+        "app_subnet_name": "[concat(parameters('projectName'), '-APP-SN')]",
+        "bastion_name": "[concat(parameters('projectName'), '-BH')]",
+        "bastion_publicIp_name": "[concat(parameters('projectName'), '-BH-PublicIP')]",
+        "bastion_subnet_addressPrefix": "[concat(parameters('virtualNetworkAddressBase'), '.128/27')]",
+        "bastion_subnet_name": "AzureBastionSubnet",
+        "dbServer_appDb_name": "[tolower(custom.cleanName(parameters('projectName')))]",
+        "dbServer_name": "[concat(parameters('projectName'), '-', parameters('environmentName'), '-DB')]",
+        "dbServer_backupRetentionDays": "[if(equals(parameters('environmentName'), 'dev'), '7', '28')]",
+        "diagnostic_blobServices_name": "Default",
+        "diagnostic_storageAccount_name": "[tolower(concat(take(custom.cleanName(parameters('projectName')), 11), 'diag', take(parameters('environmentName'), 4), 'sa'))]",
+        "gateway_backendAddressPool_name": "webserver",
+        "gateway_backendHttpSettingsCollection_name": "[concat(parameters('projectName'), '-HTTP-settings')]",
+        "gateway_frontendIPConfigurations_name": "appGwPublicFrontendIp",
+        "gateway_frontendPort_http_name": "http_port",
+        "gateway_frontendPort_https_name": "https_port",
+        "gateway_gatewayIPConfigurations_name": "appGatewayIpConfig",
+        "gateway_httpListeners_http_name": "HTTP-listener",
+        "gateway_httpListeners_https_name": "HTTPS-listener",
+        "gateway_name": "[concat(parameters('projectName'), '-GW')]",
+        "gateway_publicIp_domainNameLabel": "[tolower(concat(parameters('projectName'), '-', parameters('environmentName')))]",
```

Comment:
DNS labels must be in lower case. Some other names are also converted to lower case by Azure (eg.: Database name)
As for adding the environment name token to each resource, I thought it was redundant with the RG name. I originally added the envionment name token only when the name was used to build a global service URL or identifier to avoid conflict with corresponding resource in other environements.
Fixed wherever possible.

Reply:
It's just for us its easier when we list all resources of the same type (i.e virtual machines) and we can see which ones are dev/staging/prod. Or for instance we see that a certain resource costs x$ from a subscription point of view we can see which environment the resource belongs too.